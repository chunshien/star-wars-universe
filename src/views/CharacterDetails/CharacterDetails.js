import React, { Fragment, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components'
import { useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import { getDetails } from '../../actions';
import { useHistory } from 'react-router-dom'

import Details from '../../components/Character/Details';
import Button from '../../components/Button/Button';
import Loading from '../../components/Loading/Loading';

const Header = styled.div`
    text-align: center;
    font-size: 36px;
    font-weight: bold;
    padding: 10px;
`;

const ButtonContainer = styled.div`
    padding: 10px 20%; 
`;

function CharacterDetails({loading, getDetails, details}) {
    const { type, id } = useParams();
    let history = useHistory()  

    useEffect(() => {
        getDetails(type + '/' + id);
    }, [type, id, getDetails]);

    const handleBack = useCallback(() => {history.goBack()}, [history]);
    
    return (
        <Fragment>
            <Header>Details</Header>
            {loading ? 
                <Loading />
                :
                <Fragment>
                    <Details 
                        details={details}
                    />
                    <ButtonContainer>
                        <Button onClick={handleBack} />
                    </ButtonContainer>
                </Fragment>
            }
        </Fragment>
    )
}

CharacterDetails.propTypes = {
    loading: PropTypes.bool,
    details: PropTypes.object,
    getDetails: PropTypes.func
};

const mapDispatchToProps = {
    getDetails: getDetails,
};

const mapStateToProps = (state) => ({
    details: state.details,
    loading: state.loading
})

export default  connect(mapStateToProps, mapDispatchToProps)(CharacterDetails);