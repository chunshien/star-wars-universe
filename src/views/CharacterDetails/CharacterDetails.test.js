import React from 'react';
import { Router } from "react-router-dom";
import renderer from 'react-test-renderer';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import createSagaMiddleware from 'redux-saga';

import CharacterDetails from './CharacterDetails';
import reducer from '../../reducers'
import rootSaga from '../../sagas'
import history from '../../services/history';

const sagaMiddleware = createSagaMiddleware();
const store = createStore(reducer, applyMiddleware(sagaMiddleware));
sagaMiddleware.run(rootSaga);

describe('View - CharacterDetails Test', () => {
    it('renders CharacterDetails without crashing', () => {
        const component = renderer.create(
            <Provider store={store}>
                <Router history={history}>
                    <CharacterDetails/>
                </Router>
            </Provider>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
})