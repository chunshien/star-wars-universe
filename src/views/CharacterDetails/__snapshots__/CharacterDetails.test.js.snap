// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View - CharacterDetails Test renders CharacterDetails without crashing 1`] = `
Array [
  <div
    className="sc-Axmtr gUqxmZ"
  >
    Details
  </div>,
  <div
    className="sc-AxjAm jfFdTm"
  />,
  <div
    className="sc-AxmLO burQtf"
  >
    <div
      className="sc-AxhUy cUamMb"
      onClick={[Function]}
    >
      Back
    </div>
  </div>,
]
`;
